Jasons-MacBook-Pro-2:django_intro jason$ cd /Users/jason/Desktop/CodingDojo/python_stack/my_environments
Jasons-MacBook-Pro-2:my_environments jason$ ls
djangoPy3Env    py3Env
Jasons-MacBook-Pro-2:my_environments jason$ source py3Env/bin/activate
(py3Env) Jasons-MacBook-Pro-2:my_environments jason$ source djangoPy3Env/bin/activate
(djangoPy3Env) Jasons-MacBook-Pro-2:my_environments jason$ cd ..
(djangoPy3Env) Jasons-MacBook-Pro-2:python_stack jason$ ls
_python         django          flask           my_environments
(djangoPy3Env) Jasons-MacBook-Pro-2:python_stack jason$ cd django/
(djangoPy3Env) Jasons-MacBook-Pro-2:django jason$ ls
django_full_stack       django_intro            django_orm              test.py
(djangoPy3Env) Jasons-MacBook-Pro-2:django jason$ cd django_orm/
(djangoPy3Env) Jasons-MacBook-Pro-2:django_orm jason$ ls
(djangoPy3Env) Jasons-MacBook-Pro-2:django_orm jason$ django-admin startproject harrypotter
(djangoPy3Env) Jasons-MacBook-Pro-2:django_orm jason$ cd harrypotter/
(djangoPy3Env) Jasons-MacBook-Pro-2:harrypotter jason$ LS
harrypotter     manage.py
(djangoPy3Env) Jasons-MacBook-Pro-2:harrypotter jason$ ls
harrypotter     manage.py
(djangoPy3Env) Jasons-MacBook-Pro-2:harrypotter jason$ mkdir apps
(djangoPy3Env) Jasons-MacBook-Pro-2:harrypotter jason$ cd apps/
(djangoPy3Env) Jasons-MacBook-Pro-2:apps jason$ touch __init__.py
(djangoPy3Env) Jasons-MacBook-Pro-2:apps jason$ pwd
/Users/jason/Desktop/CodingDojo/python_stack/django/django_orm/harrypotter/apps
(djangoPy3Env) Jasons-MacBook-Pro-2:apps jason$ python ../manage.py startapp hogwart_app
(djangoPy3Env) Jasons-MacBook-Pro-2:apps jason$ ls
__init__.py     hogwart_app
(djangoPy3Env) Jasons-MacBook-Pro-2:apps jason$ cd ..
(djangoPy3Env) Jasons-MacBook-Pro-2:harrypotter jason$ ls
apps            harrypotter     manage.py
(djangoPy3Env) Jasons-MacBook-Pro-2:harrypotter jason$ python manage.py runserver
Performing system checks...

Unhandled exception in thread started by <function check_errors.<locals>.wrapper at 0x1092819d8>
Traceback (most recent call last):
  File "/Users/jason/Desktop/CodingDojo/python_stack/my_environments/djangoPy3Env/lib/python3.7/site-packages/django/utils/autoreload.py", line 226, in wrapper
    fn(*args, **kwargs)
  File "/Users/jason/Desktop/CodingDojo/python_stack/my_environments/djangoPy3Env/lib/python3.7/site-packages/django/core/management/commands/runserver.py", line 121, in inner_run
    self.check(display_num_errors=True)
  File "/Users/jason/Desktop/CodingDojo/python_stack/my_environments/djangoPy3Env/lib/python3.7/site-packages/django/core/management/base.py", line 385, in check
    include_deployment_checks=include_deployment_checks,
  File "/Users/jason/Desktop/CodingDojo/python_stack/my_environments/djangoPy3Env/lib/python3.7/site-packages/django/core/management/base.py", line 372, in _run_checks
    return checks.run_checks(**kwargs)
  File "/Users/jason/Desktop/CodingDojo/python_stack/my_environments/djangoPy3Env/lib/python3.7/site-packages/django/core/checks/registry.py", line 81, in run_checks
    new_errors = check(app_configs=app_configs)
  File "/Users/jason/Desktop/CodingDojo/python_stack/my_environments/djangoPy3Env/lib/python3.7/site-packages/django/core/checks/urls.py", line 14, in check_url_config
    return check_resolver(resolver)
  File "/Users/jason/Desktop/CodingDojo/python_stack/my_environments/djangoPy3Env/lib/python3.7/site-packages/django/core/checks/urls.py", line 24, in check_resolver
    for pattern in resolver.url_patterns:
  File "/Users/jason/Desktop/CodingDojo/python_stack/my_environments/djangoPy3Env/lib/python3.7/site-packages/django/utils/functional.py", line 35, in __get__
    res = instance.__dict__[self.name] = self.func(instance)
  File "/Users/jason/Desktop/CodingDojo/python_stack/my_environments/djangoPy3Env/lib/python3.7/site-packages/django/urls/resolvers.py", line 310, in url_patterns
    patterns = getattr(self.urlconf_module, "urlpatterns", self.urlconf_module)
  File "/Users/jason/Desktop/CodingDojo/python_stack/my_environments/djangoPy3Env/lib/python3.7/site-packages/django/utils/functional.py", line 35, in __get__
    res = instance.__dict__[self.name] = self.func(instance)
  File "/Users/jason/Desktop/CodingDojo/python_stack/my_environments/djangoPy3Env/lib/python3.7/site-packages/django/urls/resolvers.py", line 303, in urlconf_module
    return import_module(self.urlconf_name)
  File "/Users/jason/Desktop/CodingDojo/python_stack/my_environments/djangoPy3Env/lib/python3.7/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1006, in _gcd_import
  File "<frozen importlib._bootstrap>", line 983, in _find_and_load
  File "<frozen importlib._bootstrap>", line 967, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 677, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 728, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/Users/jason/Desktop/CodingDojo/python_stack/django/django_orm/harrypotter/harrypotter/urls.py", line 20, in <module>
    url(r'^$', include('apps.hogwarts_apps.urls')),
  File "/Users/jason/Desktop/CodingDojo/python_stack/my_environments/djangoPy3Env/lib/python3.7/site-packages/django/conf/urls/__init__.py", line 50, in include
    urlconf_module = import_module(urlconf_module)
  File "/Users/jason/Desktop/CodingDojo/python_stack/my_environments/djangoPy3Env/lib/python3.7/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1006, in _gcd_import
  File "<frozen importlib._bootstrap>", line 983, in _find_and_load
  File "<frozen importlib._bootstrap>", line 953, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "<frozen importlib._bootstrap>", line 1006, in _gcd_import
  File "<frozen importlib._bootstrap>", line 983, in _find_and_load
  File "<frozen importlib._bootstrap>", line 965, in _find_and_load_unlocked
ModuleNotFoundError: No module named 'apps.hogwarts_apps'
^[[A^[[A^[[A^[
^[[A^X^C(djangoPy3Env) Jasons-MacBook-Pro-2:harrypotter jason$ python manage.py runserver
Performing system checks...

Unhandled exception in thread started by <function check_errors.<locals>.wrapper at 0x1057329d8>
Traceback (most recent call last):
  File "/Users/jason/Desktop/CodingDojo/python_stack/my_environments/djangoPy3Env/lib/python3.7/site-packages/django/utils/autoreload.py", line 226, in wrapper
    fn(*args, **kwargs)
  File "/Users/jason/Desktop/CodingDojo/python_stack/my_environments/djangoPy3Env/lib/python3.7/site-packages/django/core/management/commands/runserver.py", line 121, in inner_run
    self.check(display_num_errors=True)
  File "/Users/jason/Desktop/CodingDojo/python_stack/my_environments/djangoPy3Env/lib/python3.7/site-packages/django/core/management/base.py", line 385, in check
    include_deployment_checks=include_deployment_checks,
  File "/Users/jason/Desktop/CodingDojo/python_stack/my_environments/djangoPy3Env/lib/python3.7/site-packages/django/core/management/base.py", line 372, in _run_checks
    return checks.run_checks(**kwargs)
  File "/Users/jason/Desktop/CodingDojo/python_stack/my_environments/djangoPy3Env/lib/python3.7/site-packages/django/core/checks/registry.py", line 81, in run_checks
    new_errors = check(app_configs=app_configs)
  File "/Users/jason/Desktop/CodingDojo/python_stack/my_environments/djangoPy3Env/lib/python3.7/site-packages/django/core/checks/urls.py", line 14, in check_url_config
    return check_resolver(resolver)
  File "/Users/jason/Desktop/CodingDojo/python_stack/my_environments/djangoPy3Env/lib/python3.7/site-packages/django/core/checks/urls.py", line 24, in check_resolver
    for pattern in resolver.url_patterns:
  File "/Users/jason/Desktop/CodingDojo/python_stack/my_environments/djangoPy3Env/lib/python3.7/site-packages/django/utils/functional.py", line 35, in __get__
    res = instance.__dict__[self.name] = self.func(instance)
  File "/Users/jason/Desktop/CodingDojo/python_stack/my_environments/djangoPy3Env/lib/python3.7/site-packages/django/urls/resolvers.py", line 310, in url_patterns
    patterns = getattr(self.urlconf_module, "urlpatterns", self.urlconf_module)
  File "/Users/jason/Desktop/CodingDojo/python_stack/my_environments/djangoPy3Env/lib/python3.7/site-packages/django/utils/functional.py", line 35, in __get__
    res = instance.__dict__[self.name] = self.func(instance)
  File "/Users/jason/Desktop/CodingDojo/python_stack/my_environments/djangoPy3Env/lib/python3.7/site-packages/django/urls/resolvers.py", line 303, in urlconf_module
    return import_module(self.urlconf_name)
  File "/Users/jason/Desktop/CodingDojo/python_stack/my_environments/djangoPy3Env/lib/python3.7/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1006, in _gcd_import
  File "<frozen importlib._bootstrap>", line 983, in _find_and_load
  File "<frozen importlib._bootstrap>", line 967, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 677, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 728, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/Users/jason/Desktop/CodingDojo/python_stack/django/django_orm/harrypotter/harrypotter/urls.py", line 20, in <module>
    url(r'^$', include('apps.hogwart_apps.urls')),
  File "/Users/jason/Desktop/CodingDojo/python_stack/my_environments/djangoPy3Env/lib/python3.7/site-packages/django/conf/urls/__init__.py", line 50, in include
    urlconf_module = import_module(urlconf_module)
  File "/Users/jason/Desktop/CodingDojo/python_stack/my_environments/djangoPy3Env/lib/python3.7/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1006, in _gcd_import
  File "<frozen importlib._bootstrap>", line 983, in _find_and_load
  File "<frozen importlib._bootstrap>", line 953, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "<frozen importlib._bootstrap>", line 1006, in _gcd_import
  File "<frozen importlib._bootstrap>", line 983, in _find_and_load
  File "<frozen importlib._bootstrap>", line 965, in _find_and_load_unlocked
ModuleNotFoundError: No module named 'apps.hogwart_apps'
^[[A^[[A^[[B^[[B^[
^C(djangoPy3Env) Jasons-MacBook-Pro-2:harrypotter jason$ python manage.py runserver
Performing system checks...

System check identified some issues:

WARNINGS:
?: (urls.W001) Your URL pattern '^$' uses include with a regex ending with a '$'. Remove the dollar from the regex to avoid problems including URLs.

System check identified 1 issue (0 silenced).

You have 13 unapplied migration(s). Your project may not work properly until you apply the migrations for app(s): admin, auth, contenttypes, sessions.
Run 'python manage.py migrate' to apply them.

September 04, 2019 - 23:07:00
Django version 1.10, using settings 'harrypotter.settings'
Starting development server at http://127.0.0.1:8000/
Quit the server with CONTROL-C.
[04/Sep/2019 23:07:07] "GET / HTTP/1.1" 200 5
Performing system checks...

System check identified some issues:

WARNINGS:
?: (urls.W001) Your URL pattern '^$' uses include with a regex ending with a '$'. Remove the dollar from the regex to avoid problems including URLs.

System check identified 1 issue (0 silenced).

You have 13 unapplied migration(s). Your project may not work properly until you apply the migrations for app(s): admin, auth, contenttypes, sessions.
Run 'python manage.py migrate' to apply them.

September 04, 2019 - 23:08:34
Django version 1.10, using settings 'harrypotter.settings'
Starting development server at http://127.0.0.1:8000/
Quit the server with CONTROL-C.
^C(djangoPy3Env) Jasons-MacBook-Pro-2:harrypotter jason$ python3 manage.py makemigrations
System check identified some issues:

WARNINGS:
?: (urls.W001) Your URL pattern '^$' uses include with a regex ending with a '$'. Remove the dollar from the regex to avoid problems including URLs.
Migrations for 'hogwart_app':
  apps/hogwart_app/migrations/0001_initial.py:
    - Create model Hogwarts
(djangoPy3Env) Jasons-MacBook-Pro-2:harrypotter jason$ python3 manage.py makemigrations
No changes detected
(djangoPy3Env) Jasons-MacBook-Pro-2:harrypotter jason$ python3 manage.py migrate
Operations to perform:
  Apply all migrations: admin, auth, contenttypes, hogwart_app, sessions
Running migrations:
  Rendering model states... DONE
  Applying contenttypes.0001_initial... OK
  Applying auth.0001_initial... OK
  Applying admin.0001_initial... OK
  Applying admin.0002_logentry_remove_auto_add... OK
  Applying contenttypes.0002_remove_content_type_name... OK
  Applying auth.0002_alter_permission_name_max_length... OK
  Applying auth.0003_alter_user_email_max_length... OK
  Applying auth.0004_alter_user_username_opts... OK
  Applying auth.0005_alter_user_last_login_null... OK
  Applying auth.0006_require_contenttypes_0002... OK
  Applying auth.0007_alter_validators_add_error_messages... OK
  Applying auth.0008_alter_user_username_max_length... OK
  Applying hogwart_app.0001_initial... OK
  Applying sessions.0001_initial... OK
(djangoPy3Env) Jasons-MacBook-Pro-2:harrypotter jason$ python manage.py shell
Python 3.7.3 (v3.7.3:ef4ec6ed12, Mar 25 2019, 16:52:21) 
Type 'copyright', 'credits' or 'license' for more information
IPython 7.8.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: Hogwarts.objects.create(name="Draco Malfoy", house="Slytherin", pet="Unknown", year=5)         
---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
<ipython-input-1-14c06b86a137> in <module>
----> 1 Hogwarts.objects.create(name="Draco Malfoy", house="Slytherin", pet="Unknown", year=5)

NameError: name 'Hogwarts' is not defined

In [2]: from apps.hogwart_app.models import *                                                          

In [3]: Hogwarts.objects.create(name="Draco Malfoy", house="Slytherin", pet="Unknown", year=5)         
Out[3]: <Hogwarts: Hogwarts object>

In [4]: Hogwarts.objects.create(name="Harry Potter", house="Griffindor", pet="Hedwig", year=5)         
Out[4]: <Hogwarts: Hogwarts object>

In [5]: one = Hogwarts.objects.get(id=1)                                                               

In [6]: one.year = 66                                                                                  

In [7]: one.save()                                                                                     

In [8]:                                                                                                
Do you really want to exit ([y]/n)? y
(djangoPy3Env) Jasons-MacBook-Pro-2:harrypotter jason$ ls
apps            db.sqlite3      harrypotter     manage.py
(djangoPy3Env) Jasons-MacBook-Pro-2:harrypotter jason$ python3 manage.py makemigration
Unknown command: 'makemigration'
Type 'manage.py help' for usage.
(djangoPy3Env) Jasons-MacBook-Pro-2:harrypotter jason$ python3 manage.py makemigrations
Migrations for 'hogwart_app':
  apps/hogwart_app/migrations/0002_users.py:
    - Create model Users
(djangoPy3Env) Jasons-MacBook-Pro-2:harrypotter jason$ python3 manage.py migrate
Operations to perform:
  Apply all migrations: admin, auth, contenttypes, hogwart_app, sessions
Running migrations:
  Rendering model states... DONE
  Applying hogwart_app.0002_users... OK
(djangoPy3Env) Jasons-MacBook-Pro-2:harrypotter jason$ python manage.py shell
Python 3.7.3 (v3.7.3:ef4ec6ed12, Mar 25 2019, 16:52:21) 
Type 'copyright', 'credits' or 'license' for more information
IPython 7.8.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: from apps.hogwart_app.models import *                                                          

In [2]: Users.objects.create(first_name="Jason", last_name="Park", email_address="jason@gmail.com", age
   ...: =25)                                                                                           
Out[2]: <Users: Users object>

In [3]: Users.objects.create(first_name="Harry", last_name="Potter", email_address="hogwart@gmail.com",
   ...:  age=25)                                                                                       
Out[3]: <Users: Users object>

In [4]: Users.objects.create(first_name="Hermione", last_name="Granger", email_address="herimone@gmail.
   ...: com", age=30)                                                                                  
Out[4]: <Users: Users object>

In [5]: allusers = Users.objects.all()                                                                 

In [6]: for x in range(10): 
   ...:     print x                                                                                    
  File "<ipython-input-6-80645b80614a>", line 2
    print x
          ^
SyntaxError: Missing parentheses in call to 'print'. Did you mean print(x)?


In [7]: for x in range(10): 
   ...:     print(x) 
   ...:                                                                                                
0
1
2
3
4
5
6
7
8
9

In [8]: allusers = Users.objects.all()                                                                 

In [9]: for user in allusers: 
   ...:     print(user.first_name, user.last_name) 
   ...:                                                                                                
Jason Park
Harry Potter
Hermione Granger

In [10]: user2 = Users.objects.get(id=2)                                                               

In [11]: print(user2)                                                                                  
Users object

In [12]: print(users2.email_address)                                                                   
---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
<ipython-input-12-8c373595527e> in <module>
----> 1 print(users2.email_address)

NameError: name 'users2' is not defined

In [13]: print(user2.email_address)                                                                    
hogwart@gmail.com

In [14]: user2.delete()                                                                                
Out[14]: (1, {'hogwart_app.Users': 1})

In [15]: user2 = Users.objects.get(id=2)                                                               
---------------------------------------------------------------------------
DoesNotExist                              Traceback (most recent call last)
<ipython-input-15-7d61083380b1> in <module>
----> 1 user2 = Users.objects.get(id=2)

~/Desktop/CodingDojo/python_stack/my_environments/djangoPy3Env/lib/python3.7/site-packages/django/db/models/manager.py in manager_method(self, *args, **kwargs)
     83         def create_method(name, method):
     84             def manager_method(self, *args, **kwargs):
---> 85                 return getattr(self.get_queryset(), name)(*args, **kwargs)
     86             manager_method.__name__ = method.__name__
     87             manager_method.__doc__ = method.__doc__

~/Desktop/CodingDojo/python_stack/my_environments/djangoPy3Env/lib/python3.7/site-packages/django/db/models/query.py in get(self, *args, **kwargs)
    383             raise self.model.DoesNotExist(
    384                 "%s matching query does not exist." %
--> 385                 self.model._meta.object_name
    386             )
    387         raise self.model.MultipleObjectsReturned(

DoesNotExist: Users matching query does not exist.

In [16]: Users.objects.values()                                                                        
Out[16]: <QuerySet [{'id': 1, 'first_name': 'Jason', 'last_name': 'Park', 'email_address': 'jason@gmail.com', 'age': 25, 'created_at': datetime.datetime(2019, 9, 5, 0, 14, 26, 584872, tzinfo=<UTC>), 'updated_at': datetime.datetime(2019, 9, 5, 0, 14, 26, 585162, tzinfo=<UTC>)}, {'id': 3, 'first_name': 'Hermione', 'last_name': 'Granger', 'email_address': 'herimone@gmail.com', 'age': 30, 'created_at': datetime.datetime(2019, 9, 5, 0, 15, 21, 32275, tzinfo=<UTC>), 'updated_at': datetime.datetime(2019, 9, 5, 0, 15, 21, 32311, tzinfo=<UTC>)}]>

In [17]: Users.objects.all().order_by("first_name")                                                    
Out[17]: <QuerySet [<Users: Users object>, <Users: Users object>]>

In [18]:                                                                                               
Do you really want to exit ([y]/n)? y
(djangoPy3Env) Jasons-MacBook-Pro-2:harrypotter jason$ cd ..
(djangoPy3Env) Jasons-MacBook-Pro-2:django_orm jason$ ls
harrypotter
(djangoPy3Env) Jasons-MacBook-Pro-2:django_orm jason$ cd ..
(djangoPy3Env) Jasons-MacBook-Pro-2:django jason$ ls
django_full_stack       django_intro            django_orm              test.py
(djangoPy3Env) Jasons-MacBook-Pro-2:django jason$ ls
django_full_stack       django_intro            django_orm              test.py
(djangoPy3Env) Jasons-MacBook-Pro-2:django jason$ cd django_o
bash: cd: django_o: No such file or directory
(djangoPy3Env) Jasons-MacBook-Pro-2:django jason$ ls
django_full_stack       django_intro            django_orm              test.py
(djangoPy3Env) Jasons-MacBook-Pro-2:django jason$ cd django_orm/
(djangoPy3Env) Jasons-MacBook-Pro-2:django_orm jason$ ls
harrypotter
(djangoPy3Env) Jasons-MacBook-Pro-2:django_orm jason$ django-admin startproject dojo_ninja
(djangoPy3Env) Jasons-MacBook-Pro-2:django_orm jason$ ls
dojo_ninja      harrypotter
(djangoPy3Env) Jasons-MacBook-Pro-2:django_orm jason$ cd dojo_ninja/
(djangoPy3Env) Jasons-MacBook-Pro-2:dojo_ninja jason$ ls
dojo_ninja      manage.py
(djangoPy3Env) Jasons-MacBook-Pro-2:dojo_ninja jason$ mkdir apps
(djangoPy3Env) Jasons-MacBook-Pro-2:dojo_ninja jason$ cd apps/
(djangoPy3Env) Jasons-MacBook-Pro-2:apps jason$ ls
(djangoPy3Env) Jasons-MacBook-Pro-2:apps jason$ python ../manage.py startapp dojoninja
(djangoPy3Env) Jasons-MacBook-Pro-2:apps jason$ ls
dojoninja
(djangoPy3Env) Jasons-MacBook-Pro-2:apps jason$ ls
dojoninja
(djangoPy3Env) Jasons-MacBook-Pro-2:apps jason$ cd ..
(djangoPy3Env) Jasons-MacBook-Pro-2:dojo_ninja jason$ ls
apps            dojo_ninja      manage.py
(djangoPy3Env) Jasons-MacBook-Pro-2:dojo_ninja jason$ python3 manage.py makemigrations
No changes detected
(djangoPy3Env) Jasons-MacBook-Pro-2:dojo_ninja jason$ python3 manage.py migrate
Operations to perform:
  Apply all migrations: admin, auth, contenttypes, sessions
Running migrations:
  Rendering model states... DONE
  Applying contenttypes.0001_initial... OK
  Applying auth.0001_initial... OK
  Applying admin.0001_initial... OK
  Applying admin.0002_logentry_remove_auto_add... OK
  Applying contenttypes.0002_remove_content_type_name... OK
  Applying auth.0002_alter_permission_name_max_length... OK
  Applying auth.0003_alter_user_email_max_length... OK
  Applying auth.0004_alter_user_username_opts... OK
  Applying auth.0005_alter_user_last_login_null... OK
  Applying auth.0006_require_contenttypes_0002... OK
  Applying auth.0007_alter_validators_add_error_messages... OK
  Applying auth.0008_alter_user_username_max_length... OK
  Applying sessions.0001_initial... OK
(djangoPy3Env) Jasons-MacBook-Pro-2:dojo_ninja jason$ python3 manage.py shell
Python 3.7.3 (v3.7.3:ef4ec6ed12, Mar 25 2019, 16:52:21) 
Type 'copyright', 'credits' or 'license' for more information
IPython 7.8.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: from apps.dojoninja.models import *                                                            
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-1-f80e73b542cf> in <module>
----> 1 from apps.dojoninja.models import *

~/Desktop/CodingDojo/python_stack/django/django_orm/dojo_ninja/apps/dojoninja/models.py in <module>
      4 
      5 
----> 6 class Dojo(models.Model):
      7     name = models.CharField(max_length=255)
      8     city = models.CharField(max_length=255)

~/Desktop/CodingDojo/python_stack/my_environments/djangoPy3Env/lib/python3.7/site-packages/django/db/models/base.py in __new__(cls, name, bases, attrs)
    111                         "Model class %s.%s doesn't declare an explicit "
    112                         "app_label and isn't in an application in "
--> 113                         "INSTALLED_APPS." % (module, name)
    114                     )
    115 

RuntimeError: Model class apps.dojoninja.models.Dojo doesn't declare an explicit app_label and isn't in an application in INSTALLED_APPS.

In [2]: ls                                                                                             
apps/       db.sqlite3  dojo_ninja/ manage.py*

In [3]:                                                                                                
Do you really want to exit ([y]/n)? y
(djangoPy3Env) Jasons-MacBook-Pro-2:dojo_ninja jason$ ls
apps            db.sqlite3      dojo_ninja      manage.py
(djangoPy3Env) Jasons-MacBook-Pro-2:dojo_ninja jason$ cd apps
(djangoPy3Env) Jasons-MacBook-Pro-2:apps jason$ ls
dojoninja
(djangoPy3Env) Jasons-MacBook-Pro-2:apps jason$ touch __init__.py
(djangoPy3Env) Jasons-MacBook-Pro-2:apps jason$ cd ..
(djangoPy3Env) Jasons-MacBook-Pro-2:dojo_ninja jason$ ls
apps            db.sqlite3      dojo_ninja      manage.py
(djangoPy3Env) Jasons-MacBook-Pro-2:dojo_ninja jason$ python3 manage.py makemigrations
Migrations for 'dojoninja':
  apps/dojoninja/migrations/0001_initial.py:
    - Create model Dojo
    - Create model Ninja
(djangoPy3Env) Jasons-MacBook-Pro-2:dojo_ninja jason$ python3 manage.py migrate
Operations to perform:
  Apply all migrations: admin, auth, contenttypes, dojoninja, sessions
Running migrations:
  Rendering model states... DONE
  Applying dojoninja.0001_initial... OK
(djangoPy3Env) Jasons-MacBook-Pro-2:dojo_ninja jason$ python3 manage.py shell
Python 3.7.3 (v3.7.3:ef4ec6ed12, Mar 25 2019, 16:52:21) 
Type 'copyright', 'credits' or 'license' for more information
IPython 7.8.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: from apps.dojoninja.models import *                                                            

In [2]: Dojo.objects.create(name="Okamula Chusati", city="Irvine", state="CA")                         
Out[2]: <Dojo: Dojo object>

In [3]: Dojo.objects.create(name="Hello from Fullerton Dojo", city="Fullerton", state="CA")            
Out[3]: <Dojo: Dojo object>

In [4]: Dojo.objects.create(name="The Best Dojo in NYC", city="New York City", state="NY")             
Out[4]: <Dojo: Dojo object>

In [5]: alldojos = Dojo.objects.all()                                                                  

In [6]: alldojos.delete()                                                                              
Out[6]: (3, {'dojoninja.Ninja': 0, 'dojoninja.Dojo': 3})

In [7]: a = Dojo.objects.get(id=1)                                                                     
---------------------------------------------------------------------------
DoesNotExist                              Traceback (most recent call last)
<ipython-input-7-092a4d6fc728> in <module>
----> 1 a = Dojo.objects.get(id=1)

~/Desktop/CodingDojo/python_stack/my_environments/djangoPy3Env/lib/python3.7/site-packages/django/db/models/manager.py in manager_method(self, *args, **kwargs)
     83         def create_method(name, method):
     84             def manager_method(self, *args, **kwargs):
---> 85                 return getattr(self.get_queryset(), name)(*args, **kwargs)
     86             manager_method.__name__ = method.__name__
     87             manager_method.__doc__ = method.__doc__

~/Desktop/CodingDojo/python_stack/my_environments/djangoPy3Env/lib/python3.7/site-packages/django/db/models/query.py in get(self, *args, **kwargs)
    383             raise self.model.DoesNotExist(
    384                 "%s matching query does not exist." %
--> 385                 self.model._meta.object_name
    386             )
    387         raise self.model.MultipleObjectsReturned(

DoesNotExist: Dojo matching query does not exist.

In [8]: Dojo.objects.values()                                                                          
Out[8]: <QuerySet []>

In [9]: Dojo.objects.create(name="Okamula Chusati", city="Irvine", state="CA")                         
Out[9]: <Dojo: Dojo object>

In [10]: Dojo.objects.create(name="Hello from Fullerton Dojo", city="Fullerton", state="CA")           
Out[10]: <Dojo: Dojo object>

In [11]: Dojo.objects.create(name="The Best Dojo in NYC", city="New York City", state="NY")            
Out[11]: <Dojo: Dojo object>

In [12]: Dojo.objects.values()                                                                         
Out[12]: <QuerySet [{'id': 4, 'name': 'Okamula Chusati', 'city': 'Irvine', 'state': 'CA'}, {'id': 5, 'name': 'Hello from Fullerton Dojo', 'city': 'Fullerton', 'state': 'CA'}, {'id': 6, 'name': 'The Best Dojo in NYC', 'city': 'New York City', 'state': 'NY'}]>

In [13]: Ninja.objects.create(first_name="Monkey", last_name="Brain", dojo=Dojo.object.get(id=4))      
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-13-2810463e4d49> in <module>
----> 1 Ninja.objects.create(first_name="Monkey", last_name="Brain", dojo=Dojo.object.get(id=4))

AttributeError: type object 'Dojo' has no attribute 'object'

In [14]: Ninja.objects.create(first_name="Monkey", last_name="Brain", dojo=Dojo.objects.get(id=4))     
Out[14]: <Ninja: Ninja object>

In [15]: Ninja.objects.create(first_name="Elephant", last_name="Fart", dojo=Dojo.objects.get(id=4))    
Out[15]: <Ninja: Ninja object>

In [16]: Ninja.objects.create(first_name="Lion", last_name="Smell", dojo=Dojo.objects.get(id=4))       
Out[16]: <Ninja: Ninja object>

In [17]: Ninja.objects.create(first_name="Pokemon", last_name="Pikachu", dojo=Dojo.objects.get(id=5))  
Out[17]: <Ninja: Ninja object>

In [18]: Ninja.objects.create(first_name="Dumb", last_name="Dumber", dojo=Dojo.objects.get(id=5))      
Out[18]: <Ninja: Ninja object>

In [19]: Ninja.objects.create(first_name="Coding", last_name="Dojo", dojo=Dojo.objects.get(id=5))      
Out[19]: <Ninja: Ninja object>

In [20]: Ninja.objects.create(first_name="Computer", last_name="Fast", dojo=Dojo.objects.get(id=6))    
Out[20]: <Ninja: Ninja object>

In [21]: Ninja.objects.create(first_name="Iam", last_name="Nobody", dojo=Dojo.objects.get(id=6))       
Out[21]: <Ninja: Ninja object>

In [22]: Ninja.objects.create(first_name="Mr.", last_name="President", dojo=Dojo.objects.get(id=6))    
Out[22]: <Ninja: Ninja object>

In [23]:                                                                                               
Do you really want to exit ([y]/n)? 
(djangoPy3Env) Jasons-MacBook-Pro-2:dojo_ninja jason$ ls
apps            db.sqlite3      dojo_ninja      manage.py
(djangoPy3Env) Jasons-MacBook-Pro-2:dojo_ninja jason$ python3 manage.py makemigrations
No changes detected
(djangoPy3Env) Jasons-MacBook-Pro-2:dojo_ninja jason$ python3 manage.py migrate
Operations to perform:
  Apply all migrations: admin, auth, contenttypes, dojoninja, sessions
Running migrations:
  No migrations to apply.
(djangoPy3Env) Jasons-MacBook-Pro-2:dojo_ninja jason$ python3 manage.py shell
Python 3.7.3 (v3.7.3:ef4ec6ed12, Mar 25 2019, 16:52:21) 
Type 'copyright', 'credits' or 'license' for more information
IPython 7.8.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: a = Dojo.objects.get(id=4)                                                                     
---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
<ipython-input-1-2b0ba91b307e> in <module>
----> 1 a = Dojo.objects.get(id=4)

NameError: name 'Dojo' is not defined

In [2]: from apps.dojoninja.models import *                                                            

In [3]: a = Dojo.objects.get(id=4)                                                                     

In [4]: print(a)                                                                                       
Dojo object

In [5]: Dojo.objects.get(id=4)                                                                         
Out[5]: Dojo object: Okamula Chusati Irvine CA dojoninja.Ninja.None (4)

In [6]: Ninja.objects.filter(dojo=Dojo.objects.get(id=4))                                              
Out[6]: <QuerySet [Ninja object: Monkey Brain Dojo object (1), Ninja object: Elephant Fart Dojo object (2), Ninja object: Lion Smell Dojo object (3)]>

In [7]: Ninja.objects.filter(dojo=Dojo.objects.get(id=4)).all()                                        
Out[7]: <QuerySet [Ninja object: Monkey Brain Dojo object (1), Ninja object: Elephant Fart Dojo object (2), Ninja object: Lion Smell Dojo object (3)]>

In [8]: first_dojo = Dojo.objects.get(id=4)                                                            

In [9]: first_dojo.ninjas                                                                              
Out[9]: <django.db.models.fields.related_descriptors.create_reverse_many_to_one_manager.<locals>.RelatedManager at 0x1082d2a58>

In [10]: first_dojo.ninjas.all()                                                                       
Out[10]: <QuerySet [Ninja object: Monkey Brain Dojo object (1), Ninja object: Elephant Fart Dojo object (2), Ninja object: Lion Smell Dojo object (3)]>

In [11]: last_dojo = Dojo.objects.get(id=6)                                                            

In [12]: last_dojo.ninjas.all()                                                                        
Out[12]: <QuerySet [Ninja object: Computer Fast Dojo object (7), Ninja object: Iam Nobody Dojo object (8), Ninja object: Mr. President Dojo object (9)]>

In [13]: last_ninja = Ninja.objects.get(id=9)                                                          

In [14]: last_ninja.dojo                                                                               
Out[14]: Dojo object: The Best Dojo in NYC New York City NY dojoninja.Ninja.None (6)

In [15]:                                                                                               
Do you really want to exit ([y]/n)? y
(djangoPy3Env) Jasons-MacBook-Pro-2:dojo_ninja jason$ python3 manage.py makemigrations
You are trying to add a non-nullable field 'desc' to dojo without a default; we can't do that (the database needs something to populate existing rows).
Please select a fix:
 1) Provide a one-off default now (will be set on all existing rows with a null value for this column)
 2) Quit, and let me add a default in models.py
Select an option: 1
Please enter the default value now, as valid Python
The datetime and django.utils.timezone modules are available, so you can do e.g. timezone.now
Type 'exit' to exit this prompt
>>> 
Please enter some code, or 'exit' (with no quotes) to exit.
>>> "old dojo"
Migrations for 'dojoninja':
  apps/dojoninja/migrations/0002_dojo_desc.py:
    - Add field desc to dojo
(djangoPy3Env) Jasons-MacBook-Pro-2:dojo_ninja jason$ python3 manage.py migrate
Operations to perform:
  Apply all migrations: admin, auth, contenttypes, dojoninja, sessions
Running migrations:
  Rendering model states... DONE
  Applying dojoninja.0002_dojo_desc... OK
(djangoPy3Env) Jasons-MacBook-Pro-2:dojo_ninja jason$ python3 manage.py shell
Python 3.7.3 (v3.7.3:ef4ec6ed12, Mar 25 2019, 16:52:21) 
Type 'copyright', 'credits' or 'license' for more information
IPython 7.8.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: from apps.dojoninja.models import *                                                            

In [2]: Dojo.objects.all()                                                                             
Out[2]: <QuerySet [Dojo object: Okamula Chusati Irvine CA dojoninja.Ninja.None (4), Dojo object: Hello from Fullerton Dojo Fullerton CA dojoninja.Ninja.None (5), Dojo object: The Best Dojo in NYC New York City NY dojoninja.Ninja.None (6)]>

In [3]: dojo1 = Dojo.objects.get(id=4)                                                                 

In [4]: print(dojo1)                                                                                   
Dojo object

In [5]: print(dojo1.desc)                                                                              
old dojo

In [6]: Dojo.objects.create(name="Weird Dojo", city="unknown", state="ER")                             
Out[6]: Dojo object: Weird Dojo unknown ER dojoninja.Ninja.None (7)

In [7]: dojo2 = Dojo.objects.get(id=7)                                                                 

In [8]: dojo2.desc                                                                                     
Out[8]: ''

In [9]:                                                                                                